modules/json/json-parser.c

```diff
@@ -258,8 +258,7 @@ json_parser_process_array(struct json_object *jso,
 {
   gint i;
 
-  msg->num_matches = 0;
-  log_msg_unset_match(msg, 0);
+  log_msg_clear_matches(msg);
```

Comment:
```
log {
  source(s_src);
  rewrite {
    set("foo bar" value('1'));
    unset-matches();
  };
  destination { file("/dev/stdout" template("$(format-json --scope nv-pairs --key 1)\n")); };
};
```

With this patch we can still query the $1 value after the `unset-matches();` is it an intended behaviour?
I would argue that after `unset-matches()` is called every $n should be empty (""), or at least that is what I expect.

Reply:
Good catch. The log_msg_values_foreach() did not filter out out-of-range matches. I'm pushing a fix for this, however I am starting to feel that it'd be much better to explicitly unset all matches explitictly, rather than trying to track the number of matches and then checking their number in all the code paths that query name-value pairs.

Might even be faster doing it only once, instead of at every log_msg_values_foreach() iteration.

Reply:
This is the fix for this: e234146b1ae44af0cdfff06e65917d40fa14c275

Reply:
I don't want to be nitpicking especially since I've approved this PR once, but this fix in https://github.com/syslog-ng/syslog-ng/commit/e234146b1ae44af0cdfff06e65917d40fa14c275 only fixes the case when a cleared match is referenced through value-pairs, e.g. through `format-json`.

If the match (e.g. `$1`) is used later as an input in a parser or rewrite rule, then it would still hold the previously (cleared) value.
Maybe this scenario is rather a programming error, e.g. the writer of an SCL block should be aware of this.
A dummy example SCL:
```
parser example-parser() {
....
  set("foo bar" value('1'));
  unset-matches();
  syslog-parser(template('$1'));
};
```

TLDR: how about using `log_msg_unset_match()` instead of `log_msg_clear_matches()`?

Reply:
In the last iteration, I have changed the approach of resetting values, so the branch now contains a patch that changes things - again.

With that said, this case has worked correctly before, '$1' was unset because of this branch in log_msg_get_match():

```
+const gchar *
+log_msg_get_match_if_set_with_type(const LogMessage *self, gint index_, gssize *value_len,
+                                   LogMessageValueType *type)
+{
+  g_assert(index_ >= 0 && index_ < LOGMSG_MAX_MATCHES);
+
+  if (index_ >= self->num_matches)
+    return NULL;
+
+  return nv_table_get_value_if_set(self->payload, match_handles[index_], value_len, type);
+}
+
```

e.g. if a match number was outside the range of num_matches we always considered it to be unset (NOTE the NULL return). This worked correctly even before. The value-pairs code path had a bug as @OverOrion noticed correctly, as it didn't use log_msg_get_match(), rather it iterated over all name-value pairs using log_msg_values_foreach(), which did not filter out $1 in this case. That's what my last patch fixed.

With all this iterations however, at the end I've decided to go back to the original concept and call log_msg_unset() all matches whenever the num_matches would change. This eliminates the need to handle matches separately from other name-value pairs in the query APIs. 

This is this patch that reverses most of the above.

commit 855e0c4174ebbba5db05d9395a2d33efb6b1cc4f
Author: Balazs Scheidler <bazsi77@gmail.com>
Date:   Tue May 3 13:48:12 2022 +0200

    logmsg: unset matches proactively

With that we are failing the cisco-parser() testcase in light, so I am now going to find that issue.